### PR TITLE
[Refacor] Replace 'if base' with 'with read_base' context manager

### PR DESCRIPTION
**Pull request by HAOCHENYE**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily get feedback. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation

1. Replace 'if base' with 'with read_base'. 
2. Add ConfigParsingError
3. make Config.to_dict return string.
4. Only convert dict and OrederedDict when setitem or setattr for ConfigDict

## Modification

Please briefly describe what modification is made in this PR.

## BC-breaking (Optional)

Does the modification introduce changes that break the backward-compatibility of the downstream repos?
If so, please describe how it breaks the compatibility and how the downstream projects should modify their code to keep compatibility with this PR.

## Use cases (Optional)

If this PR introduces a new feature, it is better to list some use cases here, and update the documentation.

## Checklist

1. Pre-commit or other linting tools are used to fix the potential lint issues.
2. The modification is covered by complete unit tests. If not, please add more unit test to ensure the correctness.
6. If the modification has potential influence on downstream projects, this PR should be tested with downstream projects, like MMDet or MMCls.
7. The documentation has been modified accordingly, like docstring or example tutorials.

### CODE REVIEW

File: mmengine/config/config.py [hunk 1]
```diff
@@ -744,14 +761,14 @@ def _file2dict(
                 predefined variables. Defaults to True.
 
         Returns:
-            Tuple[dict, str]: Variables dictionary and text of Config.
+            Tuple[dict, str]: Variables dictionary and text of Config.i
```
Comment (zhouzaida) [inline mmengine/config/config.py:764]:
```suggestion
            Tuple[dict, str]: Variables dictionary and text of Config.
```

File: mmengine/config/utils.py [hunk 2]
```diff
@@ -42,6 +42,13 @@
 PKG2PROJECT = MODULE2PACKAGE
 
 
+class ConfigParsingError(RuntimeError):
+
+    def __init__(self, message: str) -> None:
+        super().__init__(message)
+        self.message = message
```
Comment (zhouzaida) [inline mmengine/config/utils.py:49]: Is it necessary to implement the `__init__` method here?

File: mmengine/config/config.py
Comment (zhouzaida) [inline mmengine/config/config.py:1]: Please update the comment in https://github.com/open-mmlab/mmengine/pull/1207/files#diff-aa8376fc15b5b802ef25189d351d2580a2b77c41f3962c87fbef546cafc2c2d7R529

**Review by mzr1996 — APPROVED** [review]

File: mmengine/config/utils.py [hunk 3]
```diff
@@ -42,6 +42,11 @@
 PKG2PROJECT = MODULE2PACKAGE
 
 
+class ConfigParsingError(RuntimeError):
+    """Raised when there is an error when parsing pure Python style config
+    files."""
```
Comment (zhouzaida) [inline mmengine/config/utils.py:47]:
```suggestion
    """Raise error when failed to parse pure Python style config files."""
```

**Review by zhouzaida — APPROVED** [review]

**Pull request merged**